app/controllers/sessions_controller.rb

```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+# sessions_controller.rb
+class SessionsController < ApplicationController
+  before_action :logged_in, only: :new
+  before_action :authenticate_user, only: %i[create]
+
+  def create
+    session[:user_id] = @user.id
+    flash[:notice] = 'Welcome Back to Articles on Rails'
+    redirect_to params[:previous_url]
+  end
+
+  def destroy
+    session[:user_id] = nil
+    flash[:notice] = 'See you soon on Rails!'
+    redirect_back fallback_location: root_path
+  end
+
+  private
+
+  def authenticate_user
+    @user = User.find_by email: params[:email]
+    flash[:error] = 'Email not found on Rails!'
```

Comment:
You should set the error message only if the validation fails.

Reply:
```ruby
before_action :find_user, only: :create
def create
    if @user&.authenticate(params[:password])
      session[:user_id] = @user.id
      flash[:notice] = 'Welcome Back to Articles on Rails'
      redirect_to params[:previous_url]
    else
      flash[:error] = 'Incorrect email or password, please try again.'
      redirect_to signin_path
    end
end
```
Linter throws two warnings, about  `Metrics/AbcSize` so I need to add the` before_action` and  use the  Safe navigation operator  `&.`